### ISSUE TITLE
Scope down marin3r Secret cluster wide access

### ISSUE THREAD

**Issue by roivaz** [3scale/marin3r#58]
## Cluster scope requirement

Marin3r's only supported install mode is cluster wide, basically because it needs to manage cluster scoped resources: 

* **discoveryservices.operator.marin3r.3scale.net**: this is the custom resource that users use to deploy an envoy discovery service and activate the injection of envoy sidecars in Pods of the selected namespaces.
* **mutatingwebhookconfigurations.admissionregistration.k8s.io**: this is the resource that the discoveryservice controller creates to enable pod injection of envoy sidecards in the pod resources of a namespace.

Due to the existence of these two resources that are managed by the controllers of the project, the operator needs to run using a ClusterRole, as a ClusterRole is the only way to interact with cluster scoped resources.

## Access to Secrets requirement

The operator uses certificates for three different purposes:

* It has the feature of delivering certificates to the envoy proxies through the the discovery service. These certificates are read from kubernetes Secret resources, thus the requirement of read access for Secrets.

* Related to the previous point, to protect certificate private keys from being exposed to any client that connects to the discovery service server, envoy proxies are required to authenticate with the discovery service server using mTLS, otherwise any client in the cluster could connect to the discovery service server and retrieve the certificates' private keys. To setup mTLS for both the discovery server and the envoy proxies that the operator manages, the operator creates server and client certificates using a self-signed certificate authority.

* A kubernetes admission webhook requires that a CA bundle is specified to validate the identity of the webhook server it connects to in order to validate/mutate admission requests. The CA certificate and the webhook server certificates for this are also stored as Secrets.

Link to documentation on how the operator manages certificates for these three requirements: https://github.com/3scale/marin3r/blob/master/docs/design/operator.md

## Install modes

An operator can be installed in [four modes](https://docs.openshift.com/container-platform/4.6/operators/understanding/olm/olm-understanding-operatorgroups.html#olm-operatorgroups-membership_olm-understanding-operatorgroups): AllNamespaces (aka cluster scope), MultiNamespace, SingleNamespace and OwnNamespace. So even if the operator only needs to read/write Secrets in a few namespaces, the AllNamespaces install mode is required to be able to manage the MutatingWebhookConfiguration cluster scoped resource. This effectively makes the operator have access to all Secrets in the cluster, with no possibility of scoping it down, as a mixed mode where cluster resources are accessed with a ClusterRole and namespaced resources are accessed with a Role is not possible, at least not when installing an operator through OLM.

## Possible solutions

* Disable all the Secret related marin3r features through a field in the DiscoveryService resource. Still, something external would need to provide the certificate for the mutating webhook. For example [cert-manager](https://cert-manager.io/) could be used as a certificate provider. The cert-manager operator would be the one having access to the Secrets in the cluster, but in the end we would be granting cluster scope access for Secrets to the operator that provides the certificates, be it Mari3nr or something else, so not sure this solves much.

* Adding a flag to filter out events from privileged namespaces (any namespace that matches openshift-* for example). The problem with this approach it's that the operator would still require the permissions to access Secrets cluster wide and the responsibility of filtering would be left to the operator itself, so in a way you would have to trust that the code is doing the right thing and filtering out events from those privileged namespaces. This is probably something worth doing anyway to reduce the number of events the controllers receive.

**Comment by roivaz** [comment]
There is a different option that we hadn't considered that could solve the problem:

* Instead of the operator itself deploying the webhook, let OLM install it as part of the operator's installation. OLM has embedded logic to provide the server certificates that webhooks require.
* Once the operator doesn't have to manage the webhook, the DiscoveryService resource can be converted to a namespaced resource, allowing the operator to be installed using namespaced mode. This way the operator would only have access to the namespace where the DiscoveryService resource is created. Only thing to take into account is that the DiscoveryService resource would need to be created in the same namespace where the envoy sidecars are injected.

**Comment by maleck13** [comment]
Having OLM deploy the webhook seems reasonable  . As the operator will be watching for CRs in namespace A it seems reasonable to add an instance of the discovery service to that namespace also if that makes the permissions simpler.

**Comment by maleck13** [comment]
One question do you know which version of OCP supports that?

**Comment by roivaz** [comment]
We have just released a new version of the operator with the proposed changes (see #60 ):

* Namespaced DiscoveryService resource
* Webhook deployed through OLM

We are right now testing it in our 3scale SaaS staging environment which is OCP 4.6, all good so far. 

We have also tested that the webhook installation through OLM also works with 4.5. We haven't done any test in older OCP clusters and I have been unable to find references to this feature outside of the 4.6 official documentation (see https://docs.openshift.com/container-platform/4.6/operators/operator_sdk/osdk-generating-csvs.html#olm-webhook-considerations_osdk-generating-csvs). This feature is also referred in these [olm docs](https://olm.operatorframework.io/docs/advanced-tasks/adding-admission-and-conversion-webhooks/) but not tied to a specific version.

Which OCP release are you targeting?
